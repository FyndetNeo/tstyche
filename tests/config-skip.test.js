import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, getFixtureUrl, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const tsconfig = {
  extends: "../../tsconfig.json",
  include: ["**/*"],
};

const fixtureUrl = getFixtureUrl("config-skip", { generated: true });

afterEach(async () => {
  await clearFixture(fixtureUrl);
});

describe("'--skip' command line option", () => {
  test("selects tests to run", async () => {
    const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--skip", "internal"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("selects test group to run", async () => {
    const testText = `import { describe, expect, test } from "tstyche";
describe("internal", () => {
  test("is string?", () => {
    expect<string>().type.toBeString();
  });
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--skip", "internal"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("overrides the '.only' run mode flag", async () => {
    const testText = `import { expect, test } from "tstyche";
test.only("internal is string?", () => {
  expect<string>().type.toBeString();
});

test.only("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test.only("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--skip", "internal"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when '--only' command line option is specified", async () => {
    const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test.only("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--only", "number", "--skip", "internal"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified before the option", async () => {
    const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["dummy", "--skip", "internal"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified after the option", async () => {
    const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixtureUrl, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--skip", "internal", "dummy"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });
});
