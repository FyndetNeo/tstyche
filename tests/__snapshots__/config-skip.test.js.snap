// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'--skip' command line option overrides the '.only' run mode flag: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  - skip internal is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests not matching 'internal' in all test files.
"
`;

exports[`'--skip' command line option selects test group to run: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  internal
    - skip is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests not matching 'internal' in all test files.
"
`;

exports[`'--skip' command line option selects tests to run: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  - skip internal is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests not matching 'internal' in all test files.
"
`;

exports[`'--skip' command line option when '--only' command line option is specified: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  - skip internal is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'number' and not matching 'internal' in all test files.
"
`;

exports[`'--skip' command line option when search string is specified after the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/dummy.test.ts
  - skip internal is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests not matching 'internal' in test files matching 'dummy'.
"
`;

exports[`'--skip' command line option when search string is specified before the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/dummy.test.ts
  - skip internal is string?
  - skip internal is number?
  + external is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests not matching 'internal' in test files matching 'dummy'.
"
`;
